services/npm/npm-downloads.service.js

```diff
@@ -10,95 +10,92 @@ const pointResponseSchema = Joi.object({
+  dt: {
+    query: 'range/1000-01-01:3000-01-01',
+    // https://github.com/npm/registry/blob/master/docs/download-counts.md#output-1
+    schema: Joi.object({
+      downloads: Joi.array()
+        .items(pointResponseSchema)
+        .required(),
+    }).required(),
+    transform: json =>
+      json.downloads
+        .map(item => item.downloads)
+        .reduce((accum, current) => accum + current),
+    messageSuffix: '',
+  },
+}
 
-  const schema = isRange ? rangeResponseSchema : pointResponseSchema
+// This hits an entirely different API from the rest of the NPM services, so
+// it does not use NpmBase.
+module.exports = class NpmDownloads extends BaseJsonService {
+  static get category() {
+    return 'downloads'
+  }
 
-  // This hits an entirely different API from the rest of the NPM services, so
-  // it does not use NpmBase.
-  return class NpmDownloads extends BaseJsonService {
-    static get name() {
-      return name
+  static get route() {
+    return {
+      base: 'npm',
+      pattern: ':interval(dw|dm|dy|dt)/:scope(@.+)?/:packageName',
     }
+  }
 
-    static get category() {
-      return 'downloads'
-    }
+  static get examples() {
+    return [
+      {
+        title: 'npm',
+        namedParams: { interval: 'dw', packageName: 'localeval' },
+        staticPreview: this.render({ interval: 'dw', downloadCount: 30000 }),
+        keywords: ['node'],
+      },
+    ]
+  }
 
-    static get route() {
-      return {
-        base,
-        pattern: ':scope(@.+)?/:packageName',
-      }
-    }
+  // For testing.
+  static get _intervalMap() {
+    return intervalMap
+  }
 
-    static get examples() {
-      return [
-        {
-          title: 'npm',
-          pattern: ':packageName',
-          namedParams: { packageName: 'localeval' },
-          staticPreview: this.render({ downloads: 30000 }),
-          keywords: ['node'],
-        },
-      ]
-    }
+  static render({ interval, downloadCount }) {
+    const { messageSuffix } = intervalMap[interval]
 
-    static render({ downloads }) {
-      return {
-        message: `${metric(downloads)}${messageSuffix}`,
-        color: downloads > 0 ? 'brightgreen' : 'red',
-      }
+    return {
+      message: `${metric(downloadCount)}${messageSuffix}`,
+      color: downloadCount > 0 ? 'brightgreen' : 'red',
     }
+  }
 
-    async handle({ scope, packageName }) {
-      const slug = scope ? `${scope}/${packageName}` : packageName
-      let { downloads } = await this._requestJson({
-        schema,
-        url: `https://api.npmjs.org/downloads/${query}/${slug}`,
-        errorMessages: { 404: 'package not found or too new' },
-      })
-      if (isRange) {
-        downloads = downloads
-          .map(item => item.downloads)
-          .reduce((accum, current) => accum + current)
-      }
-      return this.constructor.render({ downloads })
-    }
+  async handle({ interval, scope, packageName }) {
+    const { query, schema, transform } = intervalMap[interval]
```

Comment:
I like using destructuring here. It makes this code particularly readable and I think a bit less error prone.

Reply:
👍 I like this a lot too